### PR TITLE
python37: init at 3.7.0b1

### PR DESCRIPTION
**Pull request by FRidh**
Beta version of Python 3.7.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

The current version builds and works. Even though, we should:
- [ ] use hash-based bytecode [PEP 552](https://www.python.org/dev/peps/pep-0552/)
- [ ] check the UTF-8 changes

### CODE REVIEW

**Comment by Mic92** [comment]
Finally we can remove many encoding work-arounds in `checkPhase`: https://docs.python.org/3.7/whatsnew/3.7.html#pep-538-legacy-c-locale-coercion

Also great for my own projects and sanity.

File: pkgs/development/interpreters/python/cpython/3.7/default.nix [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    substituteInPlace configure --replace '-Wl,-stack_size,1000000' ' '
+  '';
+
+  patches = [
+    ./no-ldconfig.patch
+  ];
+
+  postPatch = ''
+    # Determinism
+    substituteInPlace "Lib/py_compile.py" --replace "source_stats['mtime']" "(1 if 'DETERMINISTIC_BUILD' in os.environ else source_stats['mtime'])"
```
Comment (lsix) [inline pkgs/development/interpreters/python/cpython/3.7/default.nix]: I did not go through [pep-0552](https://www.python.org/dev/peps/pep-0552/) intensively, but could not use the fact that it offers the possibility to have timestamp-less .pyc files and replace this hack?

something like:

```bash
substituteInPlace "Lib/py_compile.py" --replace "invalidation_mode=PycInvalidationMode.TIMESTAMP):" "invalidation_mode=PycInvalidationMode.CHECKED_HASH):"
```

and remove `DETERMINISTIC_BUILD` env var ? (this wold require some testing…)

Comment (FRidh) [reply pkgs/development/interpreters/python/cpython/3.7/default.nix]: yep, that's on the todo list. We will need to modify some of this patching.

Comment (lsix) [reply pkgs/development/interpreters/python/cpython/3.7/default.nix]:
```bash
substituteInPlace "Lib/py_compile.py" --replace "invalidation_mode=PycInvalidationMode.TIMESTAMP):" "invalidation_mode=(PycInvalidationMode.CHECKED_HASH if 'DETERMINISTIC_BUILD' in os.environ else PycInvalidationMode.TIMESTAMP)):"
```
 would probably be better

**Comment by jtojnar** [comment]
Would not this still need `C.UTF-8` locale? https://github.com/NixOS/nixpkgs/issues/20192

**Comment by FRidh** [comment]
> Would not this still need C.UTF-8 locale? #20192

https://docs.python.org/3.7/whatsnew/3.7.html#pep-540-forced-utf-8-runtime-mode

> When in UTF-8 mode, CPython ignores the locale settings, and uses the UTF-8 encoding by default. The error handlers for sys.stdin and sys.stdout streams are set to surrogateescape.
>
> The forced UTF-8 mode can be used to change the text handling behavior in an embedded Python interpreter without changing the locale settings of an embedding application.
>
> The UTF-8 mode is enabled by default when the locale is “C”. See PEP 538: Legacy C Locale Coercion for details.

So, yes. While your workstation typically has a UTF locale configured, this is or may not be the case on say servers or inside services.

**Comment by xeji** [comment]
@FRidh I believe this is obsolete since  7261a66d713aabf30249535bfa53e5d205c4423e

**Comment by FRidh** [comment]
Yep, closing.

**Pull request closed without merging**